display_list/geometry/dl_region.h

```diff
@@ -28,36 +38,111 @@ class DlRegion {
+    SpanBuffer(SpanBuffer&& m);
+
+    void reserve(size_t capacity);
+    size_t capacity() const { return capacity_; }
+
+    SpanChunkHandle storeChunk(const Span* begin, const Span* end);
+    size_t getChunkSize(SpanChunkHandle handle) const;
+    void getSpans(SpanChunkHandle handle,
+                  const DlRegion::Span*& begin,
+                  const DlRegion::Span*& end) const;
+
+    ~SpanBuffer();
+
+   private:
+    size_t capacity_ = 0;
+    size_t size_ = 0;
+
+    // Spans for the region chunks. First span in each chunk contains the
+    // chunk size.
+    Span* spans_ = nullptr;
+  };
+
+  DlRegion();
+
+  bool isEmpty() const { return lines_.empty(); }
+  bool isComplex() const;
+
   struct SpanLine {
     int32_t top;
     int32_t bottom;
-    SpanVec* spans;
-
-    void insertSpan(int32_t left, int32_t right);
-    bool spansEqual(const SpanLine& l2) const;
+    SpanChunkHandle chunk_handle;
   };
 
-  typedef std::vector<SpanLine> LineVec;
+  void addRects(const std::vector<SkIRect>& rects);
 
-  std::vector<SpanLine> lines_;
-  std::vector<SpanVec*> spanvec_pool_;
-
-  void insertLine(size_t position, SpanLine line);
-  LineVec::iterator removeLine(LineVec::iterator position);
+  void appendLine(int32_t top,
+                  int32_t bottom,
+                  const Span* begin,
+                  const Span* end);
+  void appendLine(int32_t top,
+                  int32_t bottom,
+                  const SpanBuffer& buffer,
+                  SpanChunkHandle handle) {
+    const Span *begin, *end;
+    buffer.getSpans(handle, begin, end);
+    appendLine(top, bottom, begin, end);
+  }
 
+  typedef std::vector<Span> SpanVec;
+  SpanLine makeLine(int32_t top, int32_t bottom, const SpanVec&);
   SpanLine makeLine(int32_t top,
                     int32_t bottom,
-                    int32_t spanLeft,
-                    int32_t spanRight);
-  SpanLine makeLine(int32_t top, int32_t bottom, const SpanVec& spans);
+                    const Span* begin,
+                    const Span* end);
+  static size_t unionLineSpans(std::vector<Span>& res,
+                               const SpanBuffer& a_buffer,
+                               SpanChunkHandle a_handle,
+                               const SpanBuffer& b_buffer,
+                               SpanChunkHandle b_handle);
+  static size_t intersectLineSpans(std::vector<Span>& res,
+                                   const SpanBuffer& a_buffer,
+                                   SpanChunkHandle a_handle,
+                                   const SpanBuffer& b_buffer,
+                                   SpanChunkHandle b_handle);
+
+  bool spansEqual(SpanLine& line, const Span* begin, const Span* end) const;
+
+  static bool spansIntersect(const Span* begin1,
+                             const Span* end1,
+                             const Span* begin2,
+                             const Span* end2);
+
+  std::vector<SpanLine> lines_;
```

Comment:
(const) vector of `const SpanLine`?

Reply:
I'm wondering how we can avoid modifying a const Region, though. Somehow when you used an iterator for these fields, it didn't seem to care that you were modifying the internals of an object that you had declared const.

Reply:
I see what you mean. It did care. I was never modifying iterator of those regions:
```cpp
  auto a_lines = a.lines_;  // This creates copy of vector
  auto b_lines = b.lines_;  // This creates copy of vector
  auto a_it = a_lines.begin(); // Iterator of local copy of a.lines_, can be modified
  auto b_it = b_lines.begin(); // Iterator of local copy of b.lines_, can be modified
```
In retrospect this is was pretty confusing naming.

If I tried to actually modify a.lines_ iterator, that would not compile because `begin()` would give me `const_iterator`.

Reply:
D'oh! I get it now...